src/java/org/apache/cassandra/repair/autorepair/AutoRepairUtils.java

```diff
@@ -447,6 +453,182 @@ public static AutoRepairHistory getHostWithLongestUnrepairTime(RepairType repair
+            Map<AbstractReplicationStrategy, List<String>> replicationStrategies = getReplicationStrategies();
+
+            // 5. Filter out repair histories who have a replica being repaired, note that this is lazy, given the stream
+            //    is completed using findFirst, it should stop as soon as the matching criteria is met.
+            finishedRepairHistories = finishedRepairHistories.filter((history) -> !hasReplicaWithOngoingRepair(history,
+                                                                                                               hostsBeingRepaired,
+                                                                                                               myId,
+                                                                                                               replicationStrategies));
+        }
+
+        // 6. Select the first (oldest lastRepairFinishTime) repair history without replicas being repaired
+        return finishedRepairHistories.findFirst().orElse(null);
+    }
+
+
+    /**
+     * @return Whether the host for the given eligibleRepairHistory has any replicas in hostsBeingRepaired.
+     * @param eligibleHistory History of node to check
+     * @param hostsBeingRepaired Hosts being repaired.
+     * @param myId Host id of this node, if the repair history is for this node, additional logging will take place.
+     * @param replicationStrategies Mapping of unique replication strategies to keyspaces having that strategy.
+     */
+    private static boolean hasReplicaWithOngoingRepair(AutoRepairHistory eligibleHistory,
+                                                       Map<InetAddressAndPort, UUID> hostsBeingRepaired,
+                                                       UUID myId,
+                                                       Map<AbstractReplicationStrategy, List<String>> replicationStrategies)
+    {
+        // If no broadcast address found for this host id in cluster metadata, just skip it, a node should always
+        // see itself in cluster metadata.
+        InetAddressAndPort eligibleBroadcastAddress = getBroadcastAddress(eligibleHistory.hostId);
+        if (eligibleBroadcastAddress == null)
+        {
+            return true;
+        }
+
+        // For each replication strategy, determine if host being repaired is a replica of the local node.
+        for (Map.Entry<AbstractReplicationStrategy, List<String>> entry : replicationStrategies.entrySet())
+        {
+            AbstractReplicationStrategy replicationStrategy = entry.getKey();
+            EndpointsByRange endpointsByRange = replicationStrategy.getRangeAddresses(ClusterMetadata.current());
+
+            // get ranges of the eligible address for the given replication strategy.
+            RangesAtEndpoint rangesAtEndpoint = StorageService.instance.getReplicas(replicationStrategy, eligibleBroadcastAddress);
```

Comment:
I suspect there might be an opportunity to optimize this further.

This method returns all ranges the given address has for the given replication (including non-primary ranges), and then for each of those ranges it finds endpoints also having that range.

It's almost a certainty that this code is currently evaluating the same endpoints multiple times.  Think I can avoid that by keeping a set of evaluated endpoints and then iterating over them individually.   Will change that.

Reply:
eh, just realizing the work done per endpoint is just a simple map existence check, any optimization involving checking set existence isn't any better, so not sure if worth changing.